src/drafty.js

```diff
@@ -1698,242 +1459,466 @@ Drafty.attrValue = function(style, data) {
-      children: []
-    });
-  });
-  spans.sort((x, y) => {
-    // start asc.
-    if (x.start != y.start) {
-      return x.start - y.start;
+  if (!Array.isArray(fmt)) {
+    if (ent.length == 1) {
+      // Handle special case when all values in fmt are 0 and fmt is skipped.
+      fmt = [{
+        at: 0,
+        len: 0,
+        key: 0
+      }];
+    } else {
+      return [];
+    }
+  }
+
+  const textLen = txt.length;
+
+  let spans = [].concat(fmt);
+  // Zero values may have been stripped. Restore them.
+  // Also ensure indexes and lengths are sane.
+  spans.forEach((s) => {
+    s.at = s.at || 0;
+    if (s.at < -1) {
+      s.at = -1;
+      s.len = 0;
+    }
+    s.len = s.len || 0;
+    if (s.len < 0) {
+      s.len = 0;
+    } else if (s.at > textLen) {
+      s.len = 0;
+    } else if (s.at + s.len > textLen) {
+      s.len = textLen - s.at;
     }
-    // break ties by end desc.
-    return y.end - x.end;
   });
-  const frst = toTree(spans);
-
-  let last = -1;
-  let offset = 0;
-  const ent_refs = [];
-  let ref_cnt = 0;
-  const toDrafty = (forest) => {
-    for (let i in forest) {
-      const span = forest[i];
-      if (span.tp && span.tp == 'QQ') {
-        last = span.end;
-        offset += span.end - span.start + 1;
-        continue;
-      }
-      if (span.end > last) {
-        preview.txt += txt.substring(last + 1, span.end + 1);
-        last = span.end;
-      }
-      const fmt = {
-        at: span.start - offset,
-        len: span.end - span.start + 1
-      };
-      if (span.tp) {
-        fmt.tp = span.tp;
-      } else if (typeof span.key != 'undefined' && (span.key in ent)) {
-        const entity = ent[span.key];
-        if (!entity) {
-          continue;
-        }
-        if (!(span.key in ent_refs)) {
-          ent_refs[span.key] = preview.ent.length;
-          preview.ent.push(entity);
-        }
-        fmt.key = ent_refs[span.key];
-      } else {
-        continue;
+
+  // Denormalize entities into spans. Create a copy of the objects to leave
+  // original Drafty object unchanged.
+  spans = spans.map((s) => {
+    let data;
+    let tp = s.tp;
+    if (!tp) {
+      s.key = s.key || 0;
+      if (ent[s.key]) {
+        data = ent[s.key].data;
+        tp = ent[s.key].tp;
       }
-      preview.fmt.push(fmt);
-      toDrafty(span.children);
     }
-  };
-  toDrafty(frst);
-  if (last + 1 < txt.length) {
-    preview.txt += txt.substring(last + 1);
+
+    // Type still not defined? Hide invalid element.
+    tp = tp || 'HD';
+
+    return {
+      tp: tp,
+      at: s.at,
+      len: s.len,
+      data: data,
+      key: s.key
+    };
+  });
+
+  // Sort spans first by start index (asc) then by length (desc).
+  spans.sort((a, b) => {
+    const diff = a.at - b.at;
+    return diff != 0 ? diff : b.len - a.len;
+  });
+
+  // Drop invalid spans and staggered spans: the second and subsequent
+  // spans when they overlap like '_first *second_ third*'.
+  let end = -2;
```

Comment:
Any specific reason for -2? Attachments?
If so please consider Number.NEGATIVE_INFINITY to make it more readable.

Reply:
Used Number.MIN_SAFE_INTEGER